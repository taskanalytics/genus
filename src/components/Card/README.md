```react|span-3
showSource: true
---
<Card width={null}>
  <Heading as='h3'>Heading</Heading>
  <Text>Anything goes.</Text>
</Card>
```

```react|span-3
showSource: true
---
<Card>
  <Heading as='h3'>Nice photo</Heading>
  <img src='https://picsum.photos/g/400/200' />
</Card>
```

```react|span-3
showSource: true
---
<Card
  bg="goodDark"
  color="white"
  shadow="good"
  actions={[{
    name: 'Setup',
    action: () => console.log('Clicked Action 1')
  }]}
>
  <Heading as='h3'>Heading</Heading>
  <Text>You can customise card visuals</Text>
</Card>
```

# With actions
```react|span-4
showSource: true
---
<Card
actions={[
  {
    name: 'Setup',
    action: () => console.log('Clicked Action 1')
  }, {
    name: 'Delete',
    action: () => console.log('Clicked Action 2'),
    type: 'destructive',
  }
]}>
  <Heading as='h3'>Heading</Heading>
  <Text>Anything goes.</Text>
</Card>
```
