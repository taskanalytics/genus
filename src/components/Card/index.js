/** @jsx jsx */
import React, { Component } from 'react'
import T from 'prop-types'
import { jsx } from '@emotion/core'
import { withTheme } from 'emotion-theming'

import Dropdown from '../Dropdown'

import {
  StyledCard,
  StyledMoreButton,
  StyledMoreIcon,
} from './styled'

const MoreIcon = () =>
  <StyledMoreIcon size={24} viewBox="0 0 24 24">
    <path d="M12,14 C10.8954305,14 10,13.1045695 10,12 C10,10.8954305 10.8954305,10 12,10 C13.1045695,10 14,10.8954305 14,12 C14,13.1045695 13.1045695,14 12,14 Z M19,14 C17.8954305,14 17,13.1045695 17,12 C17,10.8954305 17.8954305,10 19,10 C20.1045695,10 21,10.8954305 21,12 C21,13.1045695 20.1045695,14 19,14 Z M5,14 C3.8954305,14 3,13.1045695 3,12 C3,10.8954305 3.8954305,10 5,10 C6.1045695,10 7,10.8954305 7,12 C7,13.1045695 6.1045695,14 5,14 Z"/>
  </StyledMoreIcon>

const Actions = withTheme(({ actions, p }) => {
  const size = 24

  return (
    <div css={{
      position: 'absolute',
      top: size,
      right: size,
      width: size,
      height: size,
      padding: 0,
    }}>
      <Dropdown
        right
        renderTrigger={props => (
          <StyledMoreButton size={24} type="button" {...props}>
            <MoreIcon size={size} />
            <span>Open actions</span>
          </StyledMoreButton>
        )}
        actions={actions} />
    </div>
  )
})

class Card extends Component {
  render () {
    const { actions, children, ...props } = this.props
    return (
      <StyledCard p={3} {...props} data-genus="Card">
        {actions && <Actions p={3} actions={actions} {...props} />}
        {children}
      </StyledCard>
    )
  }
}

Actions.propTypes = {
  actions: T.array,
  p: T.oneOfType([T.number, T.string]),
}

Card.propTypes = {
  actions: T.array,
  children: T.node,
}

Card.defaultProps = {
  bg: 'white',
  width: 1,
}

export default Card
