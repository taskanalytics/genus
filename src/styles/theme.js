import { rgba } from 'polished'
import { colors } from './colors'

const theme = {
  breakpoints: [
    '32em', '48em', '64em',
  ],
  space: [
    0, 6, 12, 18, 24, 48,
  ],
  radius: 4,
  fontSizes: [
    12, 16, 18, 24, 36, 72
  ],
  colors: {
    black: 'rgb(17, 17, 17)',
    blue: '#07c',
    text: 'rgb(45, 54, 68)',
    white: '#fff',
    muted: 'rgba(45, 54, 68, 0.6)',
    primary: colors.primary,
    primaryShaded: colors.primaryShaded,
    good: rgba(140, 215, 150, 1),
    goodDark: '#3BBE77',
    headsUp: rgba(255, 212, 0, 1),
    headsUpDark: '#FFA707',
    trouble: rgba(255, 90, 20, 1),
    troubleDark: '#FF5915',
  },
}

export default theme
