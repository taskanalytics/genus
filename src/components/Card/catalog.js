import Card from '.'
import Heading from '../Heading'
import Text from '../Text'

export default {
  title: 'Card',
  path: '/components/card',
  imports: { Card, Heading, Text },
  component: require('./README.md'),
}
