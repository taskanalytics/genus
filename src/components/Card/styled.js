import glamorous from 'glamorous'
import { Box } from '../Grid'
import theme from '../../styles/theme'

export const StyledCard = glamorous(Box)({
  width: '100%',
  alignItems: 'center',
  backgroundColor: theme.colors.white,
  borderRadius: theme.radius*2,
  transition: 'box-shadow .2s',
  position: 'relative',
  '&:hover': {
    boxShadow: '0 2px 6px 0 rgba(0,0,0,.1)',
  },
})

export const StyledMoreButton = glamorous.button({
  height: '24px',
  width: '24px',
  border: 'none',
  backgroundColor: 'transparent',
  padding: 0,
  fill: theme.colors.muted,
  cursor: 'pointer',
  '& span': {
    display: 'none',
  },
  '&:hover': {
    fill: theme.colors.primary,
  },
  '&:active,&:focus': {
    outline: 'none',
  }
})

export const StyledMoreIcon = glamorous.svg({
  height: '24px',
  width: '24px',
})
