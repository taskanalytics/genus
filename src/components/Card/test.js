import React from 'react'
import { ThemeProvider } from 'emotion-theming'
import { render, fireEvent, waitForElement } from '@testing-library/react'

import theme from '../../styles/theme'
import Card from './index.js'

const renderComponent = props =>
  render(
    <ThemeProvider theme={theme}>
      <Card {...props} />
    </ThemeProvider>,
  )

describe('<Card> component', () => {
  it('renders a heading', () => {
    const { container } = renderComponent({ children: <span>Text</span> })
    const node = container.firstChild
    expect(node).toMatchSnapshot()
    expect(node.tagName).toBe('DIV')
    expect(node.textContent).toBe('Text')
  })

  it('renders actions when clicked', async () => {
    const { container } = renderComponent({
      children: <span>Text</span>,
      actions: [{ name: 'Test' }],
    })
    const node = container.firstChild
    expect(node).toMatchSnapshot()
    expect(node.tagName).toBe('DIV')
    const actionsContainer = node.querySelector('[data-genus="Dropdown"]')
    const button = node.querySelector('div button')
    expect(button.firstChild.tagName).toBe('svg')

    fireEvent.click(button)

    const actions = await waitForElement(() =>
      actionsContainer.querySelectorAll('div[open]>button'),
    )

    expect(actions[1].textContent).toBe('Test')
  })
})
