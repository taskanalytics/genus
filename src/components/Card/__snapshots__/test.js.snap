// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card> component renders a heading 1`] = `
<div
  class="css-19ep89c"
  data-genus="Card"
>
  <span>
    Text
  </span>
</div>
`;

exports[`<Card> component renders actions when clicked 1`] = `
<div
  class="css-19ep89c"
  data-genus="Card"
>
  <div
    class="css-aaukym"
  >
    <div
      class="css-1r3m7kt"
      data-genus="Dropdown"
    >
      <button
        class="css-105ttwj"
        title="Open actions"
        type="button"
      >
        <svg
          height="24px"
          viewBox="0 0 24 24"
          width="24px"
        >
          <path
            d="M12,14 C10.8954305,14 10,13.1045695 10,12 C10,10.8954305 10.8954305,10 12,10 C13.1045695,10 14,10.8954305 14,12 C14,13.1045695 13.1045695,14 12,14 Z M19,14 C17.8954305,14 17,13.1045695 17,12 C17,10.8954305 17.8954305,10 19,10 C20.1045695,10 21,10.8954305 21,12 C21,13.1045695 20.1045695,14 19,14 Z M5,14 C3.8954305,14 3,13.1045695 3,12 C3,10.8954305 3.8954305,10 5,10 C6.1045695,10 7,10.8954305 7,12 C7,13.1045695 6.1045695,14 5,14 Z"
          />
        </svg>
        <span>
          Open actions
        </span>
      </button>
      <div
        class="css-8q8itd"
        data-genus="Box"
      >
        <button
          class="css-ms7u4k"
          data-genus="Button"
          type="button"
        >
          X
        </button>
        <button
          class="css-nmwv8r"
        >
          Test
        </button>
      </div>
    </div>
  </div>
  <span>
    Text
  </span>
</div>
`;
